backend/lib/utils/terminals/terminalBootstrap.js

```diff
@@ -0,0 +1,405 @@
+  const subsets = [
+    {
+      addresses: [
+        {
+          ip
+        }
+      ],
+      ports: [
+        {
+          port: 443,
+          protocol: 'TCP'
+        }
+      ]
+    }
+  ]
+
+  const body = toEndpointResource({ name, namespace, subsets, ownerReferences })
+  const client = coreClient.ns(namespace).endpoints
+
+  return replaceResource({ client, name, body })
+}
+
+async function replaceServiceKubeApiServer ({ name = TERMINAL_KUBE_APISERVER, coreClient, namespace, externalName = undefined, ownerReferences, clusterIP = 'None' }) {
+  let type
+  if (externalName) {
+    type = 'ExternalName'
+  }
+
+  const spec = {
+    clusterIP,
+    ports: [
+      {
+        port: 443,
+        protocol: 'TCP',
+        targetPort: 443
+      }
+    ],
+    type, // optional
+    externalName // optional
+  }
+
+  const body = toServiceResource({ name, namespace, spec, ownerReferences })
+  const client = coreClient.ns(namespace).services
+
+  return replaceResource({ client, name, body })
+}
+
+async function handleSeed (seed) {
+  const name = seed.metadata.name
+  const namespace = 'garden'
+
+  logger.debug(`replacing resources on seed ${name} for webterminals`)
+  const coreClient = kubernetes.core()
+  const gardenClient = kubernetes.garden()
+
+  // now make sure a a browser-trusted certificate is presented for the kube-apiserver
+  const isShootedSeed = await shoots.exists({ gardenClient, namespace, name })
+  if (isShootedSeed) {
+    await ensureTrustedCertForShootApiServer({ gardenClient, coreClient, namespace, name })
+  } else {
+    await ensureTrustedCertForSeedApiServer({ coreClient, seed })
+  }
+}
+
+async function handleShoot (shoot, cb) {
+  const name = shoot.metadata.name
+  const namespace = shoot.metadata.namespace
+  logger.debug(`replacing shoot's apiserver ingress ${name} for webterminals`)
+  const coreClient = kubernetes.core()
+  const gardenClient = kubernetes.garden()
+
+  await ensureTrustedCertForShootApiServer({ gardenClient, coreClient, namespace, name })
+}
+
+/*
+  Currently the kube apiserver of a shoot presents a certificate signed by a custom CA root certificate which is usually not trusted by a browser.
+  As for the web terminals, the frontend client opens a websocket connection directly to the kube apiserver. This requires a browser trusted certificate.
+  Preferred, the gardener provides the kube-apiserver a browser trusted certificate using the `--tls-sni-cert-key` argument.
+  Until this is the case we need to workaround this by creating an ingress (e.g. with the respective certmanager annotations) so that a proper certificate is presented for the kube-apiserver.
+  https://github.com/gardener/gardener/issues/1413
+*/
+async function ensureTrustedCertForShootApiServer ({ gardenClient, coreClient, namespace, name }) {
+  const shootResource = await shoots.read({ gardenClient, namespace, name })
```

Comment:
do we need to read it again?

Reply:
for the handleShoot case yes as the shoot resource that we have at hand is potentially old (because of queueing)